### PR TITLE
DG-1893 Support append/remove/replace for Tags

### PR DESCRIPTION
**Pull request by hr2904**
## Change description

> Added support for sending explicit actions for tag append/update/remove
New payload schema looks like 
with query param `overrideClassifications=false`
example URL :  `https://preview.atlan.com/api/meta/entity/bulk/setClassifications?overrideClassifications=false`
When override is `false`, the appends/updates/removes array will be processed instead of the default classifications array.

```
{
    "guidHeaderMap": {
        "d762a137-2aa7-4916-ae67-cd8ba86d692c": {
            "guid": "d762a137-2aa7-4916-ae67-cd8ba86d692c",
            "typeName": "Schema",
            "attributes": {
                "name": "MYSQL_RDS_FIVETRAN_METADATA_API_CHINOOK",
                "qualifiedName": "default/snowflake/1703211402/FIVETRAN/MYSQL_RDS_FIVETRAN_METADATA_API_CHINOOK"
            },
            "appendClassifications": [
                {
                    "typeName": "KsEjhLQyJe1IK3WImU66T5",
                    "entityGuid": "d762a137-2aa7-4916-ae67-cd8ba86d692c",
                    "propagate": false,
                    "restrictPropagationThroughLineage": false,
                    "removePropagationsOnEntityDelete": false
                }
            ],
            "updateClassifications": [
                {
                    "typeName": "x7rq3qNdVO64e3RecLTkOr",
                    "entityGuid": "d762a137-2aa7-4916-ae67-cd8ba86d692c",
                    "propagate": false,
                    "restrictPropagationThroughLineage": true,
                    "removePropagationsOnEntityDelete": false
                }
            ],
            "removeClassifications": [
                {
                    "typeName": "lJE8o8eUyD3gP63CQl8s85",
                    "entityGuid": "d762a137-2aa7-4916-ae67-cd8ba86d692c",
                    "propagate": false,
                    "restrictPropagationThroughLineage": false,
                    "removePropagationsOnEntityDelete": false
                }
            ],
            "classifications": [
                {
                    "typeName": "FkVPzxQreHyRlrCbJTbdEQ",
                    "entityGuid": "d762a137-2aa7-4916-ae67-cd8ba86d692c",
                    "propagate": false,
                    "restrictPropagationThroughLineage": false,
                    "removePropagationsOnEntityDelete": false
                }
            ]
        }
    }
}
```
## Type of change
- [ ] New feature (adds functionality)

## Related issues

> Fix [Ticket](https://atlanhq.atlassian.net/browse/DG-1893) 
> [TestCases](https://atlanhq.atlassian.net/wiki/x/LoBsHw)

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable
](https://atlanhq.atlassian.net/wiki/x/LoBsHw)

### CODE REVIEW

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java [hunk 1]
```diff
@@ -236,6 +242,49 @@ public void setClassifications(Map<String, AtlasEntityHeader> map) throws AtlasB
             RequestContext.get().setDelayTagNotifications(false);
         }
 
+        private Map<String, List<AtlasClassification>> validateAndTransfer(AtlasEntityHeader incomingEntityHeader, AtlasEntityHeader entityToBeChanged) throws AtlasBaseException {
+            Map<String, List<AtlasClassification>> operationListMap = new HashMap<>();
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java:246]: Add null validation on input parameters here

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java [hunk 2]
```diff
@@ -236,6 +242,49 @@ public void setClassifications(Map<String, AtlasEntityHeader> map) throws AtlasB
...
+                    .map(this::generateClassificationComparisonKey)
+                    .collect(Collectors.toSet());
+
+            Set<String> diff = requiredClassificationKeys.stream()
+                    .filter(key -> !preExistingClassificationKeys.contains(key))
+                    .collect(Collectors.toSet());
+
+            if (!diff.isEmpty()) {
+                String firstTypeName = diff.iterator().next().split("\\|")[1];
+                throw new AtlasBaseException(AtlasErrorCode.CLASSIFICATION_NOT_ASSOCIATED_WITH_ENTITY, firstTypeName);
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java:268]: Consider collecting all missing classifications and log all classifications

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java [hunk 3]
```diff
@@ -236,6 +242,49 @@ public void setClassifications(Map<String, AtlasEntityHeader> map) throws AtlasB
             RequestContext.get().setDelayTagNotifications(false);
         }
 
+        private Map<String, List<AtlasClassification>> validateAndTransfer(AtlasEntityHeader incomingEntityHeader, AtlasEntityHeader entityToBeChanged) throws AtlasBaseException {
+            Map<String, List<AtlasClassification>> operationListMap = new HashMap<>();
+
+            Set<String> requiredClassificationKeys = Stream.concat(
+                            Optional.ofNullable(incomingEntityHeader.getRemoveClassifications()).orElse(Collections.emptyList()).stream(),
+                            Optional.ofNullable(incomingEntityHeader.getUpdateClassifications()).orElse(Collections.emptyList()).stream()
```
Comment (PRATHAM2002-DS) [inline repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java:250]: I will suggest to handle null explicitly, you can check with Nikhil once though

File: repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java (hunk 2 above)
Comment (hr2904) [reply repository/src/main/java/org/apache/atlas/repository/store/graph/v2/ClassificationAssociator.java:268]: @nikhilbonte21 Should we send all the issue inducing classifications in 1 go in the error? The usual error thrown by this method's sibling method is 1 classification per request.

**Review by nikhilbonte21 — APPROVED** [review]
Added a minor comment regarding logging

**Review by nikhilbonte21 — APPROVED** [review]

**Pull request merged**